### PR TITLE
Prepare for release

### LINKED ISSUE #57: Publish to crates.io
We should publish the crate and its associated applications to `crates.io` (potentially bringing llamacord etc into a GitHub organization, too). Here's what I think's blocking:

- [x] Move `ggml-rs` to its own publishable crate, add safety warnings
- [x] Fix warnings
- [x] Add deny public docs, document everything
- [x] Get all pending PRs in

Anything else I'm missing, @setzer22 ?

### PR DESCRIPTION
**Pull request by philpax**
Mostly housekeeping.

Closes #57.

Still to do:

- [x] Document `llama-rs`
- [x] #62 
- [x] #18 
- [x] #80
- [x] #48
- [x] #44
- [x] Fix performance on Apple Silicon

### CODE REVIEW

**Comment by philpax** [comment]
This is good to go aside from the AS fix (need to test on work laptop).

Something I noticed is that `InferenceParameters` contains `play_back_previous_tokens` (only relevant to `inference_with_prompt`) and `inference_with_prompt` takes `maximum_token_count` as a parameter - we should consider moving these into a specific struct for parameters for `inference_with_prompt` only, like
```rust
    pub fn inference_with_prompt<E: std::error::Error + 'static>(
        &mut self,
        model: &Model,
        vocab: &Vocabulary,
        inference_params: &InferenceParameters,
        inference_with_prompt_params: &InferenceWithPromptParameters,
        prompt: &str,
        rng: &mut impl rand::Rng,
        callback: impl Fn(&str) -> Result<(), E>,
    ) -> Result<InferenceStats, InferenceError> {
}
```

We could also potentially move the `&mut impl rand::Rng` into `InferenceParameters`. I'm not sure if this is a good idea yet, but it should let us simplify the number of arguments some more.

Is there a context in which you would use the `Vocabulary` of a model without the model itself? I'm leaning towards moving that into `Model`, too.

**Comment by KerfuffleV2** [comment]
> We could also potentially move the `&mut impl rand::Rng` into `InferenceParameters`. 

You'd have to add a lifetime to the struct and maybe it would need to be mutable also, so probably not worth it.

**Comment by hhamud** [comment]
I'm on apple silicon and I haven't felt any performance issues versus the original implementation of llama.cpp in C. What are the performance issues?

File: llama-rs/src/lib.rs [hunk 1]
```diff
@@ -98,14 +105,39 @@ impl InferenceSession {
             .saturating_sub(self.params.repetition_penalty_last_n)..]
     }
 }
+impl Clone for InferenceSession {
+    fn clone(&self) -> Self {
+        let context = ggml::Context::init(self.memory_size);
+        let memory_k = context.new_tensor_1d(
+            self.memory_k.get_type(),
+            self.memory_k.get_ne()[0].try_into().unwrap(),
```
Comment (setzer22) [inline llama-rs/src/lib.rs:113]: there's `nelements()` too which might be a bit more succint

**Review by setzer22 — APPROVED** [review]
Thanks a lot for the changes! :)  Everything looks good, there's just a very minor nitpick comment from my end.

Also, why does `ggml.rs` appear as a removed file? The code seems to still be using it, and I don't see where that code has moved to :thinking:

**Comment by philpax** [comment]
> Thanks a lot for the changes! :) Everything looks good, there's just a very minor nitpick comment from my end.
> 
> Also, why does `ggml.rs` appear as a removed file? The code seems to still be using it, and I don't see where that code has moved to 🤔

It's moved to `ggml/src/lib.rs`. Not sure why Git can't detect the move.

**Review by setzer22 — APPROVED** [review]
LGTM! :smile:

**Pull request merged**